core/log.h

```diff
@@ -0,0 +1,186 @@
+#pragma once
+
+#include <sstream>
+
+#if ANDROID
+#include <android/log.h>
+#else
+#include <iostream>
+#include <ctime>
+#endif
+
+#define ANSI_COLOR_RED     "\x1b[31m"
+#define ANSI_COLOR_YELLOW  "\x1b[33m"
+#define ANSI_COLOR_GRAY    "\x1b[37m"
+#define ANSI_COLOR_RESET   "\x1b[0m"
+
+#ifndef WINDOWS
+// Remove path and extract only filename.
+#define __FILENAME__ (__builtin_strrchr(__FILE__, '/') ? __builtin_strrchr(__FILE__, '/') + 1 : __FILE__)
+#else
+#define __FILENAME__  __FILE__
+#endif
+
+// For release builds, don't show debug printfs, and just discard it into the NullStream.
+class NullStream
+{
+public:
+    template<typename TPrintable>
+    NullStream &operator<<(TPrintable const &)
+    {
+        /* no-op */
+        static NullStream nothing;
+        return nothing;
+    }
+};
+
+#if DEBUG
+#define LogDebug() LogDebugDetailed(__FILENAME__, __LINE__)
+#else
+#define LogDebug() NullStream()
+#endif
+
+
+#define LogInfo() LogInfoDetailed(__FILENAME__, __LINE__)
+#define LogWarn() LogWarnDetailed(__FILENAME__, __LINE__)
+#define LogErr() LogErrDetailed(__FILENAME__, __LINE__)
+
+
+namespace dronecore {
+
+class LogDetailed
+{
+public:
+    LogDetailed(const char *filename, int filenumber) : _s(),
+        _caller_filename(filename),
+        _caller_filenumber(filenumber)
+    {}
+
+    template <typename T>
+    LogDetailed &operator << (const T &x)
+    {
+        _s << x;
+        return *this;
+    }
+
+    virtual ~LogDetailed()
+    {
+#if ANDROID
+        switch (_log_level) {
+            case LogLevel::Debug:
+                __android_log_print(ANDROID_LOG_DEBUG, "DroneCore", "%s", _s.str().c_str());
+                break;
+            case LogLevel::Info:
+                __android_log_print(ANDROID_LOG_INFO, "DroneCore", "%s", _s.str().c_str());
+                break;
+            case LogLevel::Warn:
+                __android_log_print(ANDROID_LOG_WARN, "DroneCore", "%s", _s.str().c_str());
+                break;
+            case LogLevel::Err:
+                __android_log_print(ANDROID_LOG_ERROR, "DroneCore", "%s", _s.str().c_str());
+                break;
+        }
+#else
+
+        // Time output taken from:
+        // https://stackoverflow.com/questions/16357999#answer-16358264
+
+
+        time_t rawtime;
+        time(&rawtime);
```

Comment:
is this using a monotonic clock? I think its important for logging and real time systems to use a monotonic clock so every part of the system is using the same time, I've seen lots of operating systems time warp at boot crashing timers, timeouts or any time dependent loop, not to mention the effects on logging and the integrity of logged data

Reply:
The thing is that `monotonic` clock is not actually the local time as I understand this:

http://en.cppreference.com/w/cpp/chrono/steady_clock
> This clock is not related to wall clock time (for example, it can be time since last reboot), and is most suitable for measuring intervals.

Reply:
If we can't get a reliable clock I guess we are stuck with this, @hamishwillee can you point this out in our docs please?, logs are not using a monotonic clock hence are susceptible to time warping

Reply:
@mrpollo unless you just want seconds since you started the executable, that's fine with a monotonic clock.

Reply:
Er, I haven't been tracking this, but would I be correct in thinking that this discussion has nothing to do with the proposed public API for logging - `dronecore::Logging`?

What I THINK - please confirm:
- this  is an API for people *developing* DroneCore to see what is going on. 
- There doesn't appear to be an actual log - everything is going out to the standard cout console? 
- There doesn't appear to be any way to control what output is displayed from a user point of view - this is just nice colour coding to help people looking at the debug output work out what is critical and what is less important?
- This is not exported, so it can't be used by some third party writing DroneCore apps - only plugin developers?
- I presume you want some docs in the "Contributing" section?
- As Julian says, the fact that the clock is monotonic not very relevant - it is very unlikely to ever overflow, and will always be local time since start of exe right?